internal/controller/etcdopstask/handler/ondemandsnapshot/ondemandsnapshot.go

```diff
@@ -0,0 +1,224 @@
+			return handler.Result{
+				Description: "Failed to get etcd object - object not found",
+				Error:       druiderr.WrapError(err, ErrGetEtcd, string(druidv1alpha1.OperationPhaseRunning), "failed to get etcd object - object not found"),
+				Requeue:     false,
+			}
+		}
+		return handler.Result{
+			Description: "Failed to get etcd object due to transient error",
+			Error:       druiderr.WrapError(err, ErrGetEtcd, string(druidv1alpha1.OperationPhaseRunning), "failed to get etcd object due to transient error"),
+			Requeue:     true,
+		}
+	}
+	if err := etcd.IsReady(); err != nil {
+		return handler.Result{
+			Description: "Etcd is not ready",
+			Error:       druiderr.WrapError(err, ErrEtcdNotReady, string(druidv1alpha1.OperationPhaseRunning), "etcd is not ready"),
+			Requeue:     false,
+		}
+	}
+	var httpScheme string
+	if tlsConfig := etcd.Spec.Backup.TLS; tlsConfig != nil {
+		httpScheme = "https"
+		etcdbrCASecret := &v1.Secret{}
+		dataKey := ptr.Deref(tlsConfig.TLSCASecretRef.DataKey, "bundle.crt")
+		if err := h.client.Get(ctx, types.NamespacedName{Namespace: etcd.Namespace, Name: tlsConfig.TLSCASecretRef.Name}, etcdbrCASecret); err != nil {
+			return handler.Result{
+				Description: "Failed to get etcdbr CA secret",
+				Error:       druiderr.WrapError(err, ErrGetCASecret, string(druidv1alpha1.OperationPhaseRunning), fmt.Sprintf("failed to get etcdbr CA secret %s/%s", etcd.Namespace, tlsConfig.TLSCASecretRef.Name)),
+				Requeue:     true,
+			}
+		}
+		certData, ok := etcdbrCASecret.Data[dataKey]
+		if !ok {
+			return handler.Result{
+				Description: "CA cert data key not found in secret",
+				Error:       druiderr.WrapError(fmt.Errorf("CA cert data key %q not found in secret %s/%s", dataKey, etcdbrCASecret.Namespace, etcdbrCASecret.Name), ErrGetEtcd, string(druidv1alpha1.OperationPhaseRunning), "CA cert data key not found in secret"),
+				Requeue:     false,
+			}
+		}
+		caCerts := x509.NewCertPool()
+		if !caCerts.AppendCertsFromPEM(certData) {
+			return handler.Result{
+				Description: "Failed to append CA certs from secret",
+				Error:       druiderr.WrapError(fmt.Errorf("failed to append CA certs from secret %s/%s", etcdbrCASecret.Namespace, etcdbrCASecret.Name), ErrGetEtcd, string(druidv1alpha1.OperationPhaseRunning), "failed to append CA certs from secret"),
+				Requeue:     false,
+			}
+		}
+		httpTransport := http.Transport{
+			TLSClientConfig: &tls.Config{
+				RootCAs:    caCerts,
+				MinVersion: tls.VersionTLS12,
+			},
+		}
+		h.httpClient = http.Client{
+			Timeout:   h.httpClient.Timeout,
+			Transport: &httpTransport,
+		}
+	} else {
+		httpScheme = "http"
+	}
+	url := fmt.Sprintf("%s://%s.%s:%d/snapshot/%s", httpScheme, druidv1alpha1.GetClientServiceName(etcd.ObjectMeta), etcd.Namespace, ptr.Deref(etcd.Spec.Backup.Port, common.DefaultPortEtcdBackupRestore), h.config.Type)
+	if ptr.Deref(h.config.IsFinal, false) {
+		url += "?final=true"
+	}
+	req, err := http.NewRequest(http.MethodPost, url, nil)
```

Comment:
Ideally the http call should be created with a cancellable context. Everytime check conditions (backup store should continue to be enabled, or etcd should still be ready). If either of these conditions are no longer satisfied then you should cancel the context.

This means that you should make the http call in a go routine and get its result on a channel. Have a select after it that either waits for the response on a responseCh or cancellation of the context.